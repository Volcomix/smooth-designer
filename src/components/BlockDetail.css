.BlockDetail {
  position: absolute;
  min-width: 288px;
  transform: translate(-50%, -50%);
  transition: var(--ease-out-transition);
  z-index: 10;
}

.BlockDetail:focus-within {
  z-index: 20;
}

.BlockDetail-card {
  box-shadow: var(--depth-1-shadow) !important;
}

.BlockDetail:focus-within .BlockDetail-card {
  box-shadow: var(--depth-3-shadow) !important;
}

.BlockDetail-card > div {
  display: flex;
}

.BlockDetail-title {
  flex: 1;
}

.BlockDetail-actions {
  display: flex;
  flex-direction: column;
}

.BlockDetail-action {
  margin: 0 0 8px 0 !important;
  opacity: 0;
}

.BlockDetail:hover .BlockDetail-action {
  opacity: 1;
}

.BlockDetail-actions svg {
  color: var(--secondary-text-color) !important;
}
