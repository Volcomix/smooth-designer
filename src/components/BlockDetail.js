//@flow
import React from 'react'
import { Card, CardTitle, CardActions } from 'material-ui/Card'
import IconButton from 'material-ui/IconButton'
import AddLink from 'material-ui/svg-icons/action/settings-ethernet'
import './BlockDetail.css'
import type { Block } from '../types'
import Sized from './Sized'
import TextField from './TextField'

export type Props = Block & {
  onNameChange: (id: string, name: string) => void,
  onSizeChange: (id: string, width: number, height: number) => void,
  onLinkStart: (id: string, mouseEvent: MouseEvent) => void,
}

class BlockDetail extends React.Component<Props> {
  nameInput: ?HTMLInputElement

  componentDidMount() {
    this.nameInput && this.nameInput.focus()
  }

  render() {
    const { name, x, y } = this.props
    return (
      <Sized
        className="BlockDetail"
        style={{ left: x, top: y }}
        onSized={this.handleSized}
      >
        <Card className="BlockDetail-card">
          <CardTitle
            className="BlockDetail-title"
            title={
              <TextField
                hintText="Name"
                value={name}
                onChange={this.handleNameChange}
                inputRef={input => (this.nameInput = input)}
              />
            }
          />
          <CardActions className="BlockDetail-actions">
            <IconButton
              className="BlockDetail-action BlockDetail-addLink"
              tooltip="Add link"
              onMouseDown={this.handleLinkStart}
            >
              <AddLink />
            </IconButton>
          </CardActions>
        </Card>
      </Sized>
    )
  }

  handleNameChange = (event: {}, newName: string) =>
    this.props.onNameChange(this.props.id, newName)

  handleSized = ({ width, height }: ClientRect) => {
    if (width !== this.props.width || height !== this.props.height) {
      this.props.onSizeChange(this.props.id, width, height)
    }
  }

  handleLinkStart = (mouseEvent: MouseEvent) =>
    this.props.onLinkStart(this.props.id, mouseEvent)
}

export default BlockDetail
