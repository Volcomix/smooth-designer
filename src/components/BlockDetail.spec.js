//@flow
import React from 'react'
import { shallow, mount } from 'enzyme'
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider'
import BlockDetail, { type Props } from './BlockDetail'

const setupProps = (): Props => ({
  id: '0',
  name: 'Block',
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  onNameChange: jest.fn(),
  onSizeChange: jest.fn(),
  onLinkStart: jest.fn(),
})

const setup = () => {
  const props = setupProps()
  const wrapper = shallow(<BlockDetail {...props} />)
  return { wrapper, props }
}

it('renders without crashing', () => {
  const { wrapper } = setup()
  expect(wrapper).toMatchSnapshot()
})

it('focus name field when the block is added', () => {
  const wrapper = mount(
    <MuiThemeProvider>
      <BlockDetail {...setupProps()} />
    </MuiThemeProvider>,
  )
  expect(document.activeElement).toBe(
    wrapper.find('TextField input').instance(),
  )
})

it('calls onNameChange when the block name changes', () => {
  const { wrapper, props } = setup()
  const cardTitleProps: { title: Object } = wrapper.find('CardTitle').props()
  shallow(<div>{cardTitleProps.title}</div>)
    .find('TextField')
    .simulate('change', {}, 'New name')
  expect(props.onNameChange).toHaveBeenCalledWith('0', 'New name')
})

it('calls onSizeChange when the block size changes', () => {
  const { wrapper, props } = setup()
  wrapper.simulate('sized', { width: 10, height: 20 })
  expect(props.onSizeChange).toHaveBeenCalledWith('0', 10, 20)
})

it('does not call onSizeChange if the size does not change', () => {
  const { wrapper, props } = setup()
  wrapper.simulate('sized', { width: 0, height: 0 })
  expect(props.onSizeChange).not.toHaveBeenCalled()
})

it('calls onLinkStart when the link button is used', () => {
  const { wrapper, props } = setup()
  wrapper
    .find('.BlockDetail-addLink')
    .simulate('mouseDown', { clientX: 10, clientY: 20 })
  expect(props.onLinkStart).toHaveBeenCalledWith('0', {
    clientX: 10,
    clientY: 20,
  })
})
