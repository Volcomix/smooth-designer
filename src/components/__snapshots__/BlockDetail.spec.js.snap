// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with a specific style if linking 1`] = `
<Sized
  className="BlockDetail BlockDetail-linking"
  onSized={[Function]}
  style={
    Object {
      "left": 10,
      "top": 20,
    }
  }
>
  <Card
    className="BlockDetail-card"
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
    onMouseUp={[Function]}
  >
    <CardTitle
      className="BlockDetail-title"
      title={
        <TextField
          hintText="Name"
          inputRef={[Function]}
          onChange={[Function]}
          value="Block"
        />
      }
    />
    <CardActions
      className="BlockDetail-actions"
    >
      <IconButton
        className="BlockDetail-action BlockDetail-addLink"
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        onMouseDown={[Function]}
        tooltip="Add link"
        tooltipPosition="bottom-center"
        touch={false}
      >
        <ActionSettingsEthernet />
      </IconButton>
    </CardActions>
  </Card>
</Sized>
`;

exports[`renders without crashing 1`] = `
<Sized
  className="BlockDetail"
  onSized={[Function]}
  style={
    Object {
      "left": 10,
      "top": 20,
    }
  }
>
  <Card
    className="BlockDetail-card"
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
    onMouseUp={[Function]}
  >
    <CardTitle
      className="BlockDetail-title"
      title={
        <TextField
          hintText="Name"
          inputRef={[Function]}
          onChange={[Function]}
          value="Block"
        />
      }
    />
    <CardActions
      className="BlockDetail-actions"
    >
      <IconButton
        className="BlockDetail-action BlockDetail-addLink"
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        onMouseDown={[Function]}
        tooltip="Add link"
        tooltipPosition="bottom-center"
        touch={false}
      >
        <ActionSettingsEthernet />
      </IconButton>
    </CardActions>
  </Card>
</Sized>
`;
